### PR TITLE
HTTP API: secrets infrastructure

### PR DESCRIPTION
**Pull request by itamarst**
Fixes https://tahoe-lafs.org/trac/tahoe-lafs/ticket/3848

Part of review should be validating this matches documented behavior in the HTTP protocol spec.

### CODE REVIEW

File: src/allmydata/storage/http_server.py [hunk 1]
```diff
@@ -26,38 +30,83 @@
...
+
+def _authorization_decorator(required_secrets):
     """
     Check the ``Authorization`` header, and (TODO: in later revision of code)
     extract ``X-Tahoe-Authorization`` headers and pass them in.
     """
+    def decorator(f):
+        @wraps(f)
+        def route(self, request, *args, **kwargs):
+            if request.requestHeaders.getRawHeaders("Authorization", [None])[0] != str(
```
Comment (meejah) [inline src/allmydata/storage/http_server.py:79]: this should probably use `timing_safe_compare` (or however we call it here) I think, or clients may be able to learn the correct swissnum.

**Review by exarkun — APPROVED** [review]
Thanks.  A few comments inline.  Please at least address the comments about docstrings, the rest are optional.  Then merge (unless you want another review).

File: docs/proposed/http-storage-node-protocol.rst [hunk 2]
```diff
@@ -369,6 +369,19 @@ The authentication *type* used is ``Tahoe-LAFS``.
 The swissnum from the NURL used to locate the storage service is used as the *credentials*.
 If credentials are not presented or the swissnum is not associated with a storage service then no storage processing is performed and the request receives an ``401 UNAUTHORIZED`` response.
 
+There are also, for some endpoints, secrets sent via ``X-Tahoe-Authorization`` headers.
+If these are:
+
+1. Missing.
```
[the remark this answers is not in the document]
Comment (itamarst) [reply docs/proposed/http-storage-node-protocol.rst:375]: I think it's correct? There are three scenarios:

1. Wrong swissnum -> 401
2. Missing/additional secrets -> 400
3. Wrong secret -> 401

[the remark this answers is not in the document]
Comment (exarkun) [reply docs/proposed/http-storage-node-protocol.rst:375]: Oh, indeed.  I mixed up swissnum with additional secrets.

Still, do we want "missing swissnum" to be a 400, like "missing other secrets" is?  I guess "401" is conventional for a missing "Authorization" header... :shrug:

[the remark this answers is not in the document]
Comment (itamarst) [reply docs/proposed/http-storage-node-protocol.rst:375]: In theory none of these error code should happen in real world.

File: src/allmydata/test/test_storage_http.py [hunk 3]
```diff
@@ -15,34 +15,178 @@
...
+
+        # Not base64.
+        with self.assertRaises(ClientSecretsException):
+            _extract_secrets(["lease-renew-secret x"], {Secrets.LEASE_RENEW})
+
+        # Wrong length lease secrets (must be 32 bytes long).
+        with self.assertRaises(ClientSecretsException):
+            _extract_secrets(["lease-renew-secret eA=="], {Secrets.LEASE_RENEW})
+        with self.assertRaises(ClientSecretsException):
+            _extract_secrets(["lease-upload-secret eA=="], {Secrets.LEASE_RENEW})
```
[the remark this answers is not in the document]
Comment (itamarst) [reply src/allmydata/test/test_storage_http.py:121]: Decided to just split it up.

**Pull request merged**